helix-core/src/main/java/org/apache/helix/task/WorkflowDispatcher.java

```diff
@@ -150,7 +149,10 @@ && isTimeout(workflowCtx.getStartTime(), workflowCfg.getTimeout())) {
     // For workflows that have already reached final states, STOP should not take into effect.
     if (!TaskConstants.FINAL_STATES.contains(workflowCtx.getWorkflowState())
         && TargetState.STOP.equals(targetState)) {
-      LOG.info("Workflow {} is marked as stopped. Workflow state is {}", workflow,
+      if (workflowCtx.getWorkflowState() == TaskState.STOPPED) {
+        return;
+      }
+      LOG.debug("Workflow {} is marked as stopped. Workflow state is {}", workflow,
           workflowCtx.getWorkflowState());
       if (isWorkflowStopped(workflowCtx, workflowCfg)) {
         workflowCtx.setWorkflowState(TaskState.STOPPED);
```

Comment:
Justification: 
Ideally the message `Workflow {} is marked as stopped` should only be logged once, same for `workflowCtx.setWorkflowState`, but it's triggered for each task in the workflow. Verified there are tons of messages saying `Workflow ABC is marked as stopped. Workflow state is STOPPED`, meaning the workflow has been stopped already.
The only other place that set state STOPPED is in `JobDispatcher` where the rest of the update logic is handled. As a result, early termination (if state == TaskState.STOPPED) makes sense here.

Reply:
@qqu0127 I suggest moving this condition into the above IF.